todo_app/tests_e2e/test_selenium.py

```diff
@@ -0,0 +1,127 @@
+import os
+import pytest
+import requests
+
+from todo_app import app, data
+from threading import Thread
+from selenium import webdriver
+from selenium.webdriver.common.keys import Keys
+
+BOARD_URL = "https://api.trello.com/1/boards/"
+BOARDS_URL = "https://api.trello.com/1/members/me/boards?fields=name,url"
+
+KEY=os.environ.get('TRELLO_API_KEY')
+TOKEN=os.environ.get('TRELLO_TOKEN') 
+
+@pytest.fixture(scope='module')
+def test_app():
+    # Create the new board & update the board id environment variable 
+    board_id = create_trello_board()
+
+    os.environ['BOARD_ID'] = board_id
+
+    # construct the new application
+    application = app.create_app()
+    # start the app in its own thread.
+    thread = Thread(target=lambda: application.run(use_reloader=False)) 
+    thread.daemon = True
+    thread.start()
+    yield app
+
+    # Tear Down
+    thread.join(1) 
+    delete_trello_board(board_id)
+
+
+def create_trello_board():
+    url = BOARD_URL
+    new_board_name = "Test To Do List"
+
+    query = {
+    'key': KEY,
+    'token': TOKEN,
+    'name': new_board_name
+    }
+
+    create_test_board_response = requests.request(
+    "POST",
+    url,
+    params=query
+    )
+
+    query = {
+   'key': KEY,
+   'token': TOKEN
+    }
+
+    get_boards_response = requests.request(
+    "GET",
+    BOARDS_URL,
+    params=query
+    )
+
+    boards = get_boards_response.json()
+
+    for board in boards:
+        if board['name'] == new_board_name:
+            return board['id']   
+
+      
+def delete_trello_board(board_id):
+    url = BOARD_URL + board_id
+
+    query = {
+    'key': KEY,
+    'token': TOKEN
+    }
+
+    response = requests.request(
+    "DELETE",
+    url,
+    params=query
+    )
+    return response
+
+@pytest.fixture(scope="module")
+def driver():
+    with webdriver.Firefox() as driver:
+        yield driver    
+
+
+def test_task_journey(driver, test_app): 
+    driver.get('http://localhost:5000/')
+
+    # Add a to do item
+    test_item = "Add test to do item"
+
+    todo_title = driver.find_element_by_name('title')
+    todo_title.clear()
+    todo_title.send_keys(test_item)
+    todo_title.send_keys(Keys.RETURN)
+
+    web_page_data =  driver.page_source
+    assert test_item in web_page_data
```

Comment:
This test fails locally for me at this stage, seemingly due to a race condition. (has the page managed to redirect before we make this assertion?) - the simplest fix might be to add a momentary delay, but a better fix is to take advantage of selenium's default waiting, e.g. by finding the element with that text explicitly, and asserting it is present (with `is not None`) rather than using `driver.page_source`
E.g. something like 
```
assert driver.find_element_by_xpath('//td[contains(., "' + test_item + '")]') is not None
```

Reply:
Changes have been made to address race condition as advised